waagent: cleanups

flokli opened the pull request:
## Description of changes
This cleans up the waagent package, fixes its shipped udev rules to work again (by moving them to `$out/etc/udev/rules.d`), and moves all necessary $PATH dependencies of the `waagent` binary to the NixOS module.

This depends on https://github.com/NixOS/nixpkgs/pull/274264 (cherry-picked into here), so the udev rule works in initrd, too.

See individual commits for details.

Supersedes #269516.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

flokli committed:
waagent: nixpkgs-fmt

The comments for the function args are redundant, we already have them in runtimeDeps.

flokli committed:
waagent: use buildPythonApplication

flokli committed:
waagent: fix description in meta

Make description a single line, and put the multi-line version in a
multiline string.

flokli committed:
waagent: fixes

 - use propagatedBuildInputs, we need `distro` at runtime.
 - move addition of runtimeDeps into makeWrapperArgs
 - move binary moving into preFixup, so buildPythonApplication can do
   its thing.
 - populate $out/etc as well, so udev rules can be found.

The udev file needs to be patched to fix the /bin/chmod reference.

flokli committed:
nixos/waagent: move runtime dependencies to systemd service

The udev rules shell out to chmod, cut and readlink, which are available
for udev rules (see https://github.com/NixOS/nixpkgs/pull/274236).

flokli commented:
I booted a VM with this, and now correctly see symlinks in `/dev/disk/azure`:

```
$ ls -la /dev/disk/azure/
total 0
drwxr-xr-x  3 root root 140 Dec 14 13:38 .
drwxr-xr-x 10 root root 200 Dec 14 13:37 ..
lrwxrwxrwx  1 root root   9 Dec 14 13:37 resource -> ../../sda
lrwxrwxrwx  1 root root  10 Dec 14 13:38 resource-part1 -> ../../sda1
lrwxrwxrwx  1 root root   9 Dec 14 13:37 root -> ../../sdb
lrwxrwxrwx  1 root root  10 Dec 14 13:37 root-part1 -> ../../sdb1
drwxr-xr-x  2 root root  60 Dec 14 13:37 scsi1

$ ls -la /dev/disk/azure/scsi1/lun10 
lrwxrwxrwx 1 root root 12 Dec 14 13:37 /dev/disk/azure/scsi1/lun10 -> ../../../sdc
```

dmytrokyrychuk commented on pkgs/applications/networking/cluster/waagent/default.nix:
```diff
@@ -1,25 +1,18 @@
 
 let
   inherit (lib) makeBinPath;
+  python = python39;
```
Any reason why `python39` is sometimes referred to directly instead of as `python` within this file?

flokli replied:
Ugh, I wanted to replace all occurences with `python`, and leave the detail it's 3.9 to that let binding - and forgot to update the references. PTAL.

flokli commented:
It seems like the nicer way than hardcoding the absolute path in the udev files is actually `services.udev.path`, which also includes some defaults.

flokli committed:
nixos/waagent: provide waagent udev rules in initrd

This should make /dev/disk/azure appear in-initrd too.

Atry commented:
The upstream `setup.py` includes a custom `install` command to install binaries and udev rules to system path
https://github.com/Azure/WALinuxAgent/blob/28345a55f9b21dae89472111635fd6e41809d958/setup.py#L343

The custom `install` command does not support PEP 517 and it is supposed to be triggered by `python setup.py install`, not `pip wheel`, according to https://github.com/Azure/WALinuxAgent/blob/28345a55f9b21dae89472111635fd6e41809d958/README.md#installation

However, `buildPythonPackage` and `buildPythonApplication` would create a wheel via `pip wheel`, which skips the custom `install` command. As a result, the binaries and udev rules are included in the wheel's site-package path unexpectedly.

I wonder if we could call `setup.py install --prefix` in `postInstall` like what we did for other PEP 517 incompatible packages like https://github.com/NixOS/nixpkgs/blob/55520e0602239cf57799cc5eebcac2712b5728eb/pkgs/tools/audio/tts/default.nix#L108

Atry commented:
I think the upstream package is designed to be installed from source via `python setup.py install`, instead of installing from a wheel.

Can we change the `format` to [other](https://github.com/NixOS/nixpkgs/blob/55520e0602239cf57799cc5eebcac2712b5728eb/pkgs/development/interpreters/python/mk-python-derivation.nix#L94) and provide a custom `installPhase`, which calls `python setup.py install --prefix $out`?

flokli commented:
> The upstream `setup.py` includes a custom `install` command to install binaries and udev rules to system path https://github.com/Azure/WALinuxAgent/blob/28345a55f9b21dae89472111635fd6e41809d958/setup.py#L343
> 
> The custom `install` command does not support PEP 517 and it is supposed to be triggered by `python setup.py install`, not `pip wheel`, according to https://github.com/Azure/WALinuxAgent/blob/28345a55f9b21dae89472111635fd6e41809d958/README.md#installation
> 
> However, `buildPythonPackage` and `buildPythonApplication` would create a wheel via `pip wheel`, which skips the custom `install` command. As a result, the binaries and udev rules are included in the wheel's site-package path unexpectedly.
> 
> I wonder if we could call `setup.py install --prefix` in `postInstall` like what we did for other PEP 517 incompatible packages like
> 
> https://github.com/NixOS/nixpkgs/blob/55520e0602239cf57799cc5eebcac2712b5728eb/pkgs/tools/audio/tts/default.nix#L108

I tried setting `format = "other"`, adding `python.setuptools` to `nativeBuildInputs` and using a custom `installPhase`:

```
 installPhase = ''
   runHook preInstall
   ${python.pythonOnBuildForHost.interpreter} setup.py install --skip-build --prefix=$out
   wrapPythonPrograms
   runHook postInstall
 '';
```

However, `install_data` tried to create `/usr`:

```
       > running install_data
       > creating /nix/store/prcq7gskwcr74hk5qhb43ffjzldwivqk-waagent-2.8.0.11
       > creating /usr
       > error: could not create '/usr': Permission denied
```

I'd like to keep using the small fixup phase, we can still improve it in a followup PR.

Atry commented:
I think it fails because `data_files` uses absolute paths, where `--prefix` does not take effect.

It might be fixed by patching `setup.py` to prepend the prefix here.

https://github.com/Azure/WALinuxAgent/blob/28345a55f9b21dae89472111635fd6e41809d958/setup.py#L304C40-L304C50

Atry commented on pkgs/applications/networking/cluster/waagent/default.nix:
```diff
@@ -29,44 +22,33 @@ python39.pkgs.buildPythonPackage rec {
+  # The binary entrypoint and udev rules are placed to the wrong place.
+  # Move them to their default location.
+  preFixup = ''
+    mv $out/${python.sitePackages}/usr/sbin $out/bin
+    rm $out/bin/waagent2.0
+    rmdir $out/${python.sitePackages}/usr
+
+    mv $out/${python.sitePackages}/etc $out/
```
```suggestion
     ${python.pythonOnBuildForHost.interpreter} setup.py install_data --install-dir=$out --root=$out
```

flokli replied:
The proposed workaround in https://github.com/NixOS/nixpkgs/blob/fd7914c96f7c006047e0154dd239aa2396478094/doc/languages-frameworks/python.section.md?plain=1#L1684-L1689 doesn't work, because there's a lot more occurences. The entire file is 372 lines. I really think having 4 lines in fixup to move things into the right location is fine.

flokli commented:
We still need to wait for https://github.com/NixOS/nixpkgs/pull/274264 to land, and rebase on top of it.

flokli commented on pkgs/applications/networking/cluster/waagent/udev-set-mode-with-MODE.patch:
```diff
@@ -0,0 +1,27 @@
+-RUN+="/bin/chmod 0444 /sys/devices/virtual/dmi/id/product_uuid"
+-
+-LABEL="product_uuid-exit"
++SUBSYSTEM=="dmi", ATTR{sys_vendor}=="Microsoft Corporation", ATTR{product_name}=="Virtual Machine", MODE="0444"
```
There's something still fishy with this, I need to debug a bit more why MODE isn't sufficient to set the permissions.

flokli replied:
`/sys/devices/virtual/dmi/id/product_uuid` is not a device by itself, so setting `MODE` doesn't work. I'll replace this with a call to coreutils.

flokli commented on pkgs/applications/networking/cluster/waagent/default.nix:
```diff
@@ -29,44 +22,33 @@ python39.pkgs.buildPythonPackage rec {
+  # The binary entrypoint and udev rules are placed to the wrong place.
+  # Move them to their default location.
+  preFixup = ''
+    mv $out/${python.sitePackages}/usr/sbin $out/bin
```
It might be waagent tries to re-execute itself with the python interpreter and fails on the bash script. I also need to take a look into that.

```
Dec 15 11:54:41 mymachine waagent[2039]:     PATH=${PATH:+':'$PATH':'}
Dec 15 11:54:41 mymachine waagent[2039]:          ^
Dec 15 11:54:41 mymachine waagent[2039]: SyntaxError: invalid syntax
Dec 15 11:54:42 mymachine waagent[2046]:   File "/nix/store/pallmwamaz2fmp515nfj98df6jdhfkc1-waagent-2.8.0.11/bin/waagent", line 2
Dec 15 11:54:42 mymachine waagent[2046]:     PATH=${PATH:+':'$PATH':'}
Dec 15 11:54:42 mymachine waagent[2046]:          ^
```

flokli replied:
Indeed waagent tries to re-execute itself, using argv0, which doesn't work with the way our wrapper works.

I replaced the move to $out/bin with a custom `makeWrapper` call, and disabled the python default binary wrapping procedures.

flokli committed:
waagent: fix re-execution

waagent re-executes itself in UpdateHandler.run_latest, even if
autoupdate is disabled.
It manually spawns a python interprever with argv0, so make sure we set
this to the right place.
PATH contains our PYTHON, and PYTHONPATH stays set, so this should
somewhat still work.

flokli commented:
I've been running this for some time now.
 - waagent doesn't crash
 - I can make use of the udev rules to format the resource disk in initrd
 - The product_uuid node in /sys has the right permissions

@Atry , @dmytrokyrychuk do you have any further remarks, or is this good to merge?

dmytrokyrychuk reviewed (approved):
I didn't test this on Azure, but the code looks good👍
